### PR TITLE
While op forward for sentimental analysis

### LINKED ISSUE #6134: fluid.layers.__all__ does not contain While, RNN classes
`While`, `RNN` classes should be exposed in `fluild.layers.__all__`

### LINKED ISSUE #6135: increment should cast value to float
The buggy code is here. Increment may pass an int not float.
https://github.com/PaddlePaddle/Paddle/blob/8ac02279f28e2e944b983793bc91da8e58a75e94/python/paddle/v2/fluid/layers.py#L1443

### LINKED ISSUE #6139: Test use plain while op to forward a sentiment analysis network
The first step of Dynamic RNN is creating a very simple RNN of the sentiment analysis data, and test it forward data is good.

The milestone is #6138.

### PR DESCRIPTION
**Pull request by reyoung**
Fix #6139

Also, Fix #6134, Fix #6135

Base on PR #6102

### CODE REVIEW

File: python/paddle/v2/fluid/tests/test_dyn_rnn.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    def test_plain_while_op(self):
+        main_program = fluid.Program()
+        startup_program = fluid.Program()
+
+        with fluid.program_guard(main_program, startup_program):
+            sentence = fluid.layers.data(
+                name='word', shape=[1], dtype='int64', lod_level=1)
+            sent_emb = fluid.layers.embedding(
+                input=sentence, size=[65535, 32], dtype='float32')
```
Comment (QiJune) [inline python/paddle/v2/fluid/tests/test_dyn_rnn.py]: size = [len(self.word_dict), 32]

File: paddle/framework/backward.cc [hunk 2]
```diff
@@ -288,12 +300,21 @@ static void CreateGradVarInBlock(
...
        ++op_index) {
     std::unordered_set<std::string> new_vars;
+    auto& ctrl_flow_ops = CtrlFlowOps();
     ForEachVarName(ops[op_index]->Outputs(),
                    [&](const std::string& grad_var_name) {
-                     if (block_desc->HasVar(grad_var_name)) {
-                       return false;
+                     if (ctrl_flow_ops.find(ops[op_index]->Type()) !=
+                         ctrl_flow_ops.end()) {
+                       if (block_desc->HasVarRecursive(grad_var_name)) {
```
Comment (QiJune) [inline paddle/framework/backward.cc]: In fact, variable in Control operator will be used in nested blocks. It's created in parent block, but used in children block. So it should be find recursively.

File: paddle/framework/backward.cc [hunk 3]
```diff
@@ -333,14 +354,25 @@ std::vector<std::unique_ptr<OpDescBind>> MakeOpGrad(
...
   const std::vector<std::string>& outputs = op_desc->OutputArgumentNames();
+
   if (AllGradInSet(outputs, *no_grad_vars)) {
-    for (const std::string& name : inputs) {
-      no_grad_vars->insert(GradVarName(name));
+    VLOG(10) << "Drop operator " << op_desc->Type();
+    // FIXME: Hack code here
+    auto& ctrl_flow_ops = CtrlFlowOps();
+    if (ctrl_flow_ops.find(op_desc->Type()) == ctrl_flow_ops.end()) {
+      // Only computational op need drop input's gradient.
```
Comment (QiJune) [inline paddle/framework/backward.cc]: In fact, the backward of Control Operator is not calculating gradient. So `no_grad_vars` does no sense to Control Operator.

File: paddle/framework/op_desc.cc [hunk 4]
```diff
@@ -466,7 +466,12 @@ DDim CompileTimeInferShapeContext::GetDim(const std::string &name) const {
   auto var = block_.FindVarRecursive(name);
   PADDLE_ENFORCE(var != nullptr, "Cannot find variable %s", name);
   try {
-    return framework::make_ddim(var->Shape());
+    auto shape = var->Shape();
+    if (shape.empty()) {
```
Comment (QiJune) [inline paddle/framework/op_desc.cc]: In which case, the shape is empty?

File: paddle/operators/tensor_array_read_write_op.cc [hunk 5]
```diff
@@ -121,10 +124,13 @@ class ReadFromArrayOp : public ArrayOp {
...
     size_t offset = GetOffset(scope, dev_ctx);
-    PADDLE_ENFORCE_LT(offset, x_array.size());
-    framework::CopyFrom(x_array[offset], dev_ctx.GetPlace(), dev_ctx,
-                        out_tensor);
-    out_tensor->set_lod(x_array[offset].lod());
+    if (offset < x_array.size()) {
+      framework::CopyFrom(x_array[offset], dev_ctx.GetPlace(), dev_ctx,
+                          out_tensor);
+      out_tensor->set_lod(x_array[offset].lod());
+    } else {
```
Comment (QiJune) [inline paddle/operators/tensor_array_read_write_op.cc]: In which case will  `offset > x_array.size()` occur? Why not throw exception here?

File: paddle/framework/op_desc.cc (hunk 4 above)
Comment (reyoung) [reply paddle/framework/op_desc.cc]: The final step of RNN network. Memory's gradient could be empty.

File: paddle/operators/tensor_array_read_write_op.cc (hunk 5 above)
Comment (reyoung) [reply paddle/operators/tensor_array_read_write_op.cc]: Final timestep, memory

**Pull request merged**